fmlloader/src/main/java/net/minecraftforge/fml/loading/moddiscovery/ModInfo.java

```diff
@@ -27,262 +27,183 @@
+        String getDescription,
+        Optional<String> getLogoFile,
+        boolean getLogoBlur,
+        Optional<URL> getUpdateURL,
+        Optional<URL> getModURL,
+
+        List<? extends ModVersion> getDependencies,
+        List<ForgeFeature.Bound> getForgeFeatures,
+        Map<String, Object> getModProperties
+) implements IModInfo, IConfigurable {
     private static final Logger LOGGER = LogUtils.getLogger();
     private static final DefaultArtifactVersion DEFAULT_VERSION = new DefaultArtifactVersion("1");
     private static final Pattern VALID_MODID = Pattern.compile("^[a-z][a-z0-9_]{1,63}$");
     private static final Pattern VALID_NAMESPACE = Pattern.compile("^[a-z][a-z0-9_.-]{1,63}$");
     private static final Pattern VALID_VERSION = Pattern.compile("^\\d+.*");
 
-    private final ModFileInfo owningFile;
-    private final String modId;
-    private final String namespace;
-    private final ArtifactVersion version;
-    private final String displayName;
-    private final String description;
-    private final Optional<String> logoFile;
-    private final boolean logoBlur;
-    private final Optional<URL> updateJSONURL;
-    private final List<? extends IModInfo.ModVersion> dependencies;
-
-    private final List<ForgeFeature.Bound> features;
-    private final Map<String,Object> properties;
-    private final IConfigurable config;
-    private final Optional<URL> modUrl;
-
-    public ModInfo(final ModFileInfo owningFile, final IConfigurable config) {
-        Optional<ModFileInfo> ownFile = Optional.ofNullable(owningFile);
-        this.owningFile = owningFile;
-        this.config = config;
-
+    public static ModInfo of(ModFileInfo owningFile, IConfigurable config) {
         // These are sourced from the mod specific [[mod]] entry
-        this.modId = config.<String>getConfigElement("modId")
+        String modId = config.<String>getConfigElement("modId")
                 .orElseThrow(() -> new InvalidModFileException("Missing modId", owningFile));
 
         // verify we have a valid modid
-        if (!VALID_MODID.matcher(this.modId).matches()) {
-            LOGGER.error(LogUtils.FATAL_MARKER, "Invalid modId found in file {} - {} does not match the standard: {}", this.owningFile.getFile().getFilePath(), this.modId, VALID_MODID.pattern());
-            throw new InvalidModFileException("Invalid modId found : " + this.modId, owningFile);
+        if (!VALID_MODID.matcher(modId).matches()) {
+            LOGGER.error(LogUtils.FATAL_MARKER, "Invalid modId found in file {} - {} does not match the standard: {}", owningFile.getFile().getFilePath(), modId, VALID_MODID.pattern());
+            throw new InvalidModFileException("Invalid modId found : " + modId, owningFile);
         }
 
-        this.namespace = config.<String>getConfigElement("namespace")
-                .orElse(this.modId);
+        String namespace = config.<String>getConfigElement("namespace")
+                .orElse(modId);
```

Comment:
Honestly if you were going through and redoing this. I would add 1 or 2 helper functions:
`get(config, key)` and `get(config, key, default)` which would make a lot of these smaller `get(config, "namespace", modId)`